src/App.tsx

```diff
@@ -51,6 +52,36 @@ const App = () => {
       .getUiSchema()
       .then((uiSchema) => dispatch(Actions.setUiSchema(uiSchema)));
   }, [schemaService]);
+  useEffect(() => {
+    //try to preserve selection when schemas change
+    const oldSelection = selection;
+    if (!oldSelection) {
+      return;
+    }
+    const newSelectedUISchemaElement =
+      oldSelection && uiSchema
+        ? getCorrespondingElement(oldSelection?.uiSchema, uiSchema)
+        : undefined;
+    const newSelectedSchemaElement =
+      oldSelection && oldSelection.schema && schema
+        ? getCorrespondingElement(oldSelection.schema, schema)
+        : undefined;
+
+    if (
+      !isPathError(newSelectedSchemaElement) &&
+      !isPathError(newSelectedUISchemaElement) &&
+      newSelectedUISchemaElement
+    ) {
+      setSelection({
+        schema: newSelectedSchemaElement,
+        uiSchema: newSelectedUISchemaElement,
+      });
+    } else {
+      //clear old selection
+      setSelection(undefined);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [uiSchema, schema]);
```

Comment:
The linter rule is correct and `selection` is missing as a dependency. By not adding it as a dependency you run into the risk that it might be outdated within the effect. [See the docs](https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects)
>make sure the array includes all values from the component scope (such as props and state) that change over time and that are used by the effect. Otherwise, your code will reference stale values from previous renders.

Now of course we don't want to have `selection` as a dependency as we don't want to trigger the effect when selection changes. However this is also not necessary as we actually don't need the `selection` at all. Whenever a state is updated based on its previous value, we need to use the function form of the setter, so in this case
```
setSelection(oldSelection => {});
```
This way we always get the latest state of `selection` without using old state values nor specifying it in the dependencies.

Reply:
Ah, that's a lot nicer <3